Enhanced save & load configuration methods for server object in PTL

RKORavi opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
Currently cPickle is used to dump configuration using save_configuration() & load_configuration() is used to load it, which does not work as intended. Pickle is slow, unsafe and is not human readable. Storing configurations in json format would be a better option.


#### Describe Your Change
PBSTestSuite's Server.save_configuration() method should save all the configurations of PBS server listed below:
Commands output:
qmgr print server
qmgr print sched
qmgr print resource
qmgr list pbshook
save hooks .CF, .PY & .HK files.
pbsnodes -av
PBS Configuration file:
pbs.conf

* Load_configuration is used to load saved above configurations

#### Link to Design Doc
<!--- If there is a design, link to it here: **[project documentation area](https://pbspro.atlassian.net/wiki/spaces/PD/pages/1156284419/Update+file+format+and+file+name+for+data+saved+using+save+configuration+method+in+PTL)** --> 
https://pbspro.atlassian.net/wiki/spaces/PD/pages/1156284419/Update+file+format+and+file+name+for+data+saved+using+save+configuration+method+in+PTL


#### Attach Test Logs or Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->

[custom_setup_logs.txt](https://github.com/PBSPro/pbspro/files/3593469/custom_setup_logs.txt)
[use_cur_setup OOB logs.txt](https://github.com/PBSPro/pbspro/files/3593470/use_cur_setup.OOB.logs.txt)


<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/PBSPro/pbspro?pullRequest=1117) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/PBSPro/pbspro?pullRequest=1117) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/PBSPro/pbspro?pullRequest=1117) it.</sub>

SudeshnaMoulik commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4247,45 +4250,71 @@ def _load_configuration(self, infile, objtype=None):
+                    qmgr_cmd = qmgr +" < "+path
+                    d = self.du.run_cmd(self.hostname, qmgr_cmd, sudo=True, as_script=True)
+                    os.unlink(path)
+                    # Load site hooks
```
As below else block is handling pbs conf loading, can you please modify the comment

ashwathraop commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4247,45 +4250,97 @@ def _load_configuration(self, infile, objtype=None):
+                                self.du.run_cmd(
+                                    self.hostname, import_py_cmd,
+                                    sudo=True, as_script=True)
+                                # os.unlink(fn)
```
Remove this line if not intended to use.

sujatapatnaik52 replied:
Similar set of comments at line 4261. Have a look at them and remove it.

ashwathraop commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4247,45 +4250,97 @@ def _load_configuration(self, infile, objtype=None):
+                                self.du.run_cmd(
+                                    self.hostname, import_cfg_cmd,
+                                    sudo=True, as_script=True)
+                                # os.unlink(fn)
```
Remove this line if not intended to use.

RKORavi commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4247,45 +4250,97 @@ def _load_configuration(self, infile, objtype=None):
+                                            self.du.run_cmd(
+                                                self.hostname, cmd,
+                                                sudo=True, as_script=True)
+                                    cmd = (qmgr +
```
I am loading hook here, i was unable to assign multiple events in one command. So, i had to run a loop of all the hook events & add them one by one.
E.g;
cgroup_hooks event = execjob_begin,execjob_epilogue,execjob_end,execjob_launch,
	execjob_attach,
	execjob_resize,
	exechost_periodic,
	exechost_startup

Running a loop of all the events here and running a qmgr command for each event.
for h_event in events:
        qmgr -c "s h hook_name event += h_event

sujatapatnaik52 replied:
It is possible to use though, you may need to use some additional string formatters - `qmgr -c "c h move_hook_1 event='movejob,queuejob,runjob'"`. If it is **possible**, try to use one command.

sujatapatnaik52 replied:
Also there is an API `create_and_import_hook` recently added. Please have a look at the same if it will be of any use here.

RKORavi replied:
@sujatapatnaik52  i have made the changes, using import_hook api now

sujatapatnaik52 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4247,45 +4250,97 @@ def _load_configuration(self, infile, objtype=None):
+                    # Load pbs.conf file
+                elif k == "pbs_conf":
+                    f_w = base64.decodestring(v)
+                    conf_path = os.path.join(os.sep, 'etc', 'pbs.conf')
```
would recommend to try `PBS_CONF_PATH` first else go for `/etc/pbs.conf`.

RKORavi replied:
@sujatapatnaik52 @agrawalravi90 I have made the changes

sujatapatnaik52 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4247,45 +4250,97 @@ def _load_configuration(self, infile, objtype=None):
+                    for hpath in conf['hooks']:
+                        h_name = hpath.split(os.sep)[-1]
+                        hook_name = h_name[:-3]
+                        hook_name = str(hook_name)
```
```suggestion
                        hook_name = str(h_name[:-3])
```

RKORavi replied:
Changed

sujatapatnaik52 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -5222,42 +5277,73 @@ def save_configuration(self, outfile, mode='a'):
+        except BaseException:
+            print("Resourcedef file not found")
+        # save pbs.conf file
+        pcfg = os.path.join(os.sep, 'etc', 'pbs.conf')
```
Same comment as above.

npadole20 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4247,45 +4250,85 @@ def _load_configuration(self, infile, objtype=None):
+                                hk_contents.splitlines())
+                            for k, val in hk_vals[0].items():
+                                if k == 'event':
+                                    events = "'%s'" % val
```
Where is the 'events' variable being used?
Also, the function Sujata is referring to is "create_import_hook".

RKORavi replied:
Removed variable events

agrawalravi90 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -5222,42 +5263,74 @@ def save_configuration(self, outfile, mode='a'):
                           'resourcedef')
-        self._save_config_file(conf, rd)
-
+        try:
```
I didn't see code to load resourcedef file?

RKORavi replied:
I have removed resourcedef from both save & load config, because i am already using print server command to save resources & then loading them back using qmgr.

agrawalravi90 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -5222,42 +5263,74 @@ def save_configuration(self, outfile, mode='a'):
+            self.logger.info("Resourcedef file not found")
+        # save pbs.conf file
+        cfg_path = os.path.join(os.sep, 'etc', 'pbs.conf')
+        with open(cfg_path, 'r') as p:
```
As i suggested earlier, can you try to encode the pbs.conf dict itself (server.pbs_conf) so you can decode it into a dict which can be passed to set_pbs_config when loading it back?

RKORavi replied:
changed

agrawalravi90 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -5222,42 +5263,74 @@ def save_configuration(self, outfile, mode='a'):
-        else:
-            conf['qmgr_print_hook'] = ret['out']
-
+        ret = self.du.run_cmd(self.hostname, [pbsnodes, '-av'])
```
I didn't see equivalent code to load pbsnodes output?

RKORavi replied:
Thanks for pointing it out, I forgot to add code to load pbsnodes

agrawalravi90 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -5222,42 +5263,74 @@ def save_configuration(self, outfile, mode='a'):
             self.logger.error('Error processing file ' + outfile)
             return False
 
         return True
```
don't we need to save sched configuration files (sched_config, holidays, etc.? )

RKORavi replied:
Right now scope of this PR is to save only server attributes & files.
e.g, print server, print sched, hook files, pbs.conf file, pbsnodes -av.

RKORavi commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4247,45 +4250,83 @@ def _load_configuration(self, infile, objtype=None):
+                        self.hostname, qmgr_cmd, sudo=True, as_script=True)
+                    os.unlink(path)
+                # Load pbs.conf file
+                elif k == "pbs_conf":
```
Changed

RKORavi commented:
> I have an over-arching question: do we really need this capability in PTL? What is this used for?

Please refer to EDD : https://pbspro.atlassian.net/wiki/spaces/PD/pages/557088789/Design+for+a+supported+way+to+change+default+setup+in+PTL?showComments=true&showCommentArea=true

RKORavi commented:
[ptl_run_logs.txt](https://github.com/PBSPro/pbspro/files/3210198/ptl_run_logs.txt)
[save_config.docx](https://github.com/PBSPro/pbspro/files/3210201/save_config.docx)

sujatapatnaik52 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4247,45 +4250,97 @@ def _load_configuration(self, infile, objtype=None):
-                conf = newconf
-
+            with open(infile, 'r') as f:
+                conf = json.load(f)
```
I see you are yet to incorporate the above comment.

sujatapatnaik52 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4247,45 +4250,69 @@ def _load_configuration(self, infile, objtype=None):
+                            cmd = qmgr + " -c 'create hook %s'" % hook_name
+                            self.du.run_cmd(
+                                self.hostname, cmd, sudo=True,
+                                as_script=True)
```
This is just a question. why do you need as_script=True? Did you try making it work without that?

sujatapatnaik52 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -5222,42 +5249,67 @@ def save_configuration(self, outfile, mode='a'):
+        with open(cfg_path, 'r') as p:
+            pbs_cfg = p.readlines()
+            config = self.utils.convert_to_dictlist(pbs_cfg)
+            print("PBS config\n ---------\n")
```
If the print statements are for debugging, you can remove them.

sujatapatnaik52 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -5222,42 +5249,67 @@ def save_configuration(self, outfile, mode='a'):
-            f.close()
-        except:
+            with open(outfile, mode) as f:
+                json.dump(conf, f)
```
same try-except suggestion for the above block as well.

agrawalravi90 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4247,45 +4250,83 @@ def _load_configuration(self, infile, objtype=None):
+                if k.startswith('qmgr_'):
+                    fd, path = tempfile.mkstemp(prefix="runcmd_")
+                    os.chmod(path, 0o700)
+                    with open(path, 'a') as f:
```
please ignore this comment and only consider my comment below for using create_temp_file()

sujatapatnaik52 commented on test/fw/bin/pbs_benchpress:
```diff
@@ -1,7 +1,7 @@
 # coding: utf-8
 
-# Copyright (C) 1994-2019 Altair Engineering, Inc.
+# Copyright (C) 1994-2018 Altair Engineering, Inc.
```
I did not get this change. This is supposed to be 2019?

sujatapatnaik52 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -466,6 +468,10 @@ def setUpClass(cls):
+        fd, path = tempfile.mkstemp(prefix="new_server_output_",
+                                    suffix=".json")
+        cls.saved_file = path
+        Server().save_configuration(path, 'w')
```
I have few questions here - 
1. The file name `setupclass_output_`, the last underscore is not needed. If I am missing something here please point me the same. 
2. Is it mandatory to do `save_configuration` before `init_servers`? I was thinking `init_server` is creating the server object and thus you can use the `save_configuration` method of the same instead of creating another `Server()` object.

sujatapatnaik52 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -1526,3 +1552,4 @@ def tearDown(self):
     @classmethod
     def tearDownClass(cls):
         cls._testMethodName = 'tearDownClass'
+        Server().load_configuration(cls.saved_file)
```
I think you can try using `cls.server` here?

sujatapatnaik52 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -483,6 +483,10 @@ def setUp(self):
+        fd, path = tempfile.mkstemp(prefix="setup_output_",
+                                    suffix=".json")
+        self.saved_file = path
+        Server().save_configuration(path, 'w')
```
you can use self.servers here?

prakashcv13 commented:
@RKORavi - if you are not working on this PR - could you please mark it [WIP] or close it.

RKORavi commented:
@agrawalravi90 @kjakkali @hirenvadalia @saritakh @sujatapatnaik52 
I have made changes to revert_to_defaults() method could you please review the PR

agrawalravi90 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -9600,6 +9713,51 @@ def import_hook(self, name, body, level=logging.INFO):
+        # Setting it to lower value to avoid the race condition at hook copy
+        srv_stat = self.status(SERVER, 'sync_mom_hookfiles_timeout')
+        try:
+            sync_val = srv_stat[0]['sync_mom_hookfiles_timeout']
```
where do you use sync_val?

RKORavi replied:
Function is no longer required since i have moved to qmgr command for getting hooks output and importing it

agrawalravi90 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -9600,6 +9713,51 @@ def import_hook(self, name, body, level=logging.INFO):
+        """
+        # sync_mom_hookfiles_timeout is 15min by default
+        # Setting it to lower value to avoid the race condition at hook copy
+        srv_stat = self.status(SERVER, 'sync_mom_hookfiles_timeout')
```
where do you use srv_stat?

RKORavi replied:
Function is no longer required since i have moved to qmgr command for getting hooks output and importing it

agrawalravi90 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -9600,6 +9713,51 @@ def import_hook(self, name, body, level=logging.INFO):
+             'input-file': rfile}
+        self.manager(MGR_CMD_IMPORT, HOOK, a, name)
+
+        os.remove(rfile)
```
you can just use self.du.rm(self.hostname, rfile) for local as well, then you won't need the if condition below

RKORavi replied:
Function is no longer required since i have moved to qmgr command for getting hooks output and importing it

agrawalravi90 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -494,7 +495,11 @@ def setUpClass(cls):
+                                   suffix=".json")
+        cls.saved_file = path
         cls.init_servers()
+        cls.server.save_configuration(path, 'w')
```
trying to understand the logic, shouldn't you be saving the configuration only when '--use-current-setup' was passed by the user?

RKORavi replied:
As per EDD system should revert to original setup after running PTL even when --use-current-setup is not provided in benchpress command

sujatapatnaik52 replied:
Since saving configuration is kind of mandatory now, I think in the `save_configuration()`, you are returning `False` so maybe you should add a check here as well to proceed only if it succeeds.

agrawalravi90 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -505,11 +510,16 @@ def setUp(self):
-        self.revert_pbsconf()
+        _, path = tempfile.mkstemp(prefix="setup_output",
+                                   suffix=".json")
+        self.saved_file = path
```
won't this overwrite the 'saved_file' path that you created in setupclass() ?

RKORavi replied:
prefix of both the files are different so, i will not overwrite file created in setupclass

agrawalravi90 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -1564,9 +1574,16 @@ def tearDown(self):
 
         for sched in self.scheds:
             self.scheds[sched].cleanup_files()
+        self.revert_server(self.server)
```
why did you add the revert operation to tearDown() ?

RKORavi replied:
Sorry my bad, added this to debug and forgot to remove.

agrawalravi90 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -1564,9 +1574,16 @@ def tearDown(self):
     @classmethod
     def tearDownClass(cls):
         cls._testMethodName = 'tearDownClass'
+        cls.server.load_configuration(cls.saved_file)
```
why are you loading configuration in tearDownClass()?

RKORavi replied:
@agrawalravi90 as per the EDD: https://pbspro.atlassian.net/wiki/spaces/PD/pages/557088789/Design+for+a+supported+way+to+change+default+setup+in+PTL
It says to load PBS everytime in teardown and teardown class. 
Goal is to revert pbs to the same state after running the tests. Even when --use-current-setup is not used

agrawalravi90 commented on test/fw/ptl/utils/plugins/ptl_test_runner.py:
```diff
@@ -509,6 +510,7 @@ def set_data(self, paramfile, testparam,
             else:
                 testparam = _f
         self.param = testparam
+        self.cur_setup = use_cur_setup
```
I like use_cur_setup more than cur_setup, I'd suggest renaming cur_setup to use_cur_setup, it's more intuitive

RKORavi replied:
Changed

RKORavi commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4247,45 +4250,83 @@ def _load_configuration(self, infile, objtype=None):
+                        f.truncate(0)
+                        f.write(cfg_vals)
+                # Load hooks
+                elif k == "hooks":
```
Missed it, I will make the changes

RKORavi commented:
@sujatapatnaik52 @agrawalravi90 should i go ahead and squash my commits, it is getting really difficult when i try to rebase and resolve the conflicts.

agrawalravi90 commented:
> @sujatapatnaik52 @agrawalravi90 should i go ahead and squash my commits, it is getting really difficult when i try to rebase and resolve the conflicts.

I'd have preferred if you hadn't rebased (and resolved conflicts) until the PR was signed off. But since you've already done this, I don't mind i you squash your commits since I'll have to review it completely anyways. But please don't rebase until you get sign-offs for future PRs.

RKORavi commented:
Attached the logs of for --use-current-setup option in pbs_benchpress.
1. Logs of running pbs_benchpress on a OOB configuration with --use-current-setup
2. Logs of running pbs_benchpress on a custom setup with --use-current-setup option

RKORavi commented:
@agrawalravi90 @sujatapatnaik52 Sorry for delay in addressing all the comments, took me a while to get a travis as passed, though i am still seeing error in centos:7 Build:santize node. I will try to fix that too.

RKORavi commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4323,49 +4327,117 @@ def _load_configuration(self, infile, objtype=None):
                         if None, load all objects in infile
         """
+        #self.unset_svr_attrib()
+        self.delete_hooks()
```
changed

RKORavi commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4323,49 +4327,117 @@ def _load_configuration(self, infile, objtype=None):
                         for, one of server, scheduler, mom or
                         if None, load all objects in infile
         """
+        #self.unset_svr_attrib()
```
removed commented line

agrawalravi90 commented:
@RKORavi please don't do force push(es) to your branch during the review phase, and do NOT do any rebase(es) until your PR is approved.

agrawalravi90 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -1474,9 +1487,15 @@ def tearDown(self):
 
         for sched in self.scheds:
             self.scheds[sched].cleanup_files()
+        self.server.del_hooks_nodes_qs()
```
Interesting, this will add some part of "revert" action to tearDown() as well. We will delete hooks, nodes, queues and reservations in this function and then load the "default" configuration, where the default could either be PTL's default, or the user's provided configuration.
Ideally, we should do the following:
- split the "revert" functions into two parts: delete and load, where delete would be common but the load function will either load the PTL's default, or the user provided configuration.

But, since this PR has been hanging around for a long time, if you don't want to do the refactoring, then I suggest that you only call "self.server.del_hooks_nodes_qs()" and "self.server.load_configuration(self.saved_file)" if "self.use_cur_setup" is True. Similarly, only call "self.server.save_configuration(path, 'w')" in setUp() if "self.use_cur_setup" is True. This will mean that when "self.use_cur_setup" is not True, PTL will behave exactly like before.

agrawalravi90 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -1474,9 +1487,15 @@ def tearDown(self):
             self.scheds[sched].cleanup_files()
+        self.server.del_hooks_nodes_qs()
+        self.server.load_configuration(self.saved_file)
+        self.du.rm(path=self.saved_file)
```
Do you really need to remove the saved configuration? Why save and delete for every test? You only need to save it once in setUp() and then just load it in tearDown() right?

RKORavi replied:
Hi Ravi,

As per the EDD : https://pbspro.atlassian.net/wiki/spaces/PD/pages/557088789/Design+for+a+supported+way+to+change+default+setup+in+PTL
- We have to bring back PBS to the same custom setup on which it started.
- Saved configuration in setup() has to be loaded in teardown() everytime.
- When --use-current-setup option is used, tests should run on given custom setup.

Instead of saving the configurations for each test case. I could save configuration the first time when setup() is called for the first time in a test suite

agrawalravi90 replied:
right, that's exactly what i was suggesting. Save it only once, don't save again if already saved, then in tearDown() you don't need to delete self.saved_file

RKORavi replied:
updated changes by adding class variable: PBSTestSuite.config_saved as mentioned in design doc.

agrawalravi90 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4187,6 +4190,27 @@ def tracejob_match(self, msg, id=None, n=50, tail=True,
+        if del_hooks:
+            self.delete_hooks()
+        if del_vdef or del_nodes:
+            self.cleanup_reservations()
```
since you also delete reservations here, please add that to the method name and docstring

agrawalravi90 replied:
or you can call the method something generic, like "delete_current_state", up to you. But do mention what all will get deleted in the docstring

RKORavi replied:
Added docstring & changed function name

agrawalravi90 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4187,6 +4190,27 @@ def tracejob_match(self, msg, id=None, n=50, tail=True,
+                           del_vdef=True):
+        """
+        Delete nodes, queues and hooks
+        """
```
please document the function's arguments in the doctring

RKORavi replied:
Added function arguments as docstring

RKORavi commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4247,45 +4250,83 @@ def _load_configuration(self, infile, objtype=None):
+                conf = json.load(f)
+            qmgr = os.path.join(self.client_conf['PBS_EXEC'],
+                                'bin', 'qmgr')
             for k, v in conf.items():
```
iteritems() depricated in python3

RKORavi commented:
Travis is taking 9 mins more after this change in PTL framework

sujatapatnaik52 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4208,46 +4241,77 @@ def _load_configuration(self, infile, objtype=None):
+                    print_svr = '\n'.join(v)
+                    with open(fpath, 'w') as f:
+                        f.write(print_svr)
+                    file_qmgr = open(fpath)
```
I am not sure why are you opening the file in the above line again? Can't this be achieved with the above `with open as f`?

RKORavi replied:
In with block i am writing contents to a file. Then I am opening the file again and passing file descriptor it to run_cmd as stdin. If I add the same thing within with block then i does not work as expected. It is unable to load the configurations.

sujatapatnaik52 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4208,46 +4241,77 @@ def _load_configuration(self, infile, objtype=None):
+                # Load server configuration
+                if k.startswith('qmgr_'):
+                    fpath = self.du.create_temp_file()
+                    print_svr = '\n'.join(v)
```
Are we just loading server here or all objects? Then I think we should update the docstring accordingly.

sujatapatnaik52 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -5158,36 +5239,121 @@ def save_configuration(self, outfile, mode='a'):
+        """
+        qmgr = os.path.join(self.client_conf['PBS_EXEC'], 'bin', 'qmgr')
+        cfg = {"hooks": ""}
+        cmd = [qmgr, '-c', 'print hook @default']
```
Just have a doubt here. @default is for specifying server right? If we do `print hook` then that should also return the same output. Why are we using @default explicitly?

RKORavi replied:
As per admin guide: @default refers to default server.

sujatapatnaik52 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -5158,36 +5239,121 @@ def save_configuration(self, outfile, mode='a'):
+            pbs_cfg_b64 = base64.encodestring(str(config[0]))
+        conf['pbs_conf'] = pbs_cfg_b64
+        # save hook files
+        hooks_str = self._save_hook_files()
```
I think you should add a check here - if `_save_hook_files()` is returning `False` then you should maybe throw an exception or so?

RKORavi replied:
Changed, returning false when saving of hook has failed.

sujatapatnaik52 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -441,11 +446,22 @@ def setUp(self):
+            _, path = tempfile.mkstemp(prefix="setup_output",
+                                       suffix=".json")
+            self.saved_file = path
+            self.server.save_configuration(path, 'w')
```
Same comment as above.

RKORavi replied:
changed

sujatapatnaik52 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4208,46 +4241,77 @@ def _load_configuration(self, infile, objtype=None):
 
+    def create_pbsnode(self, node_name, attrs):
+        """
+        Create node in PBS with given attributes
```
I have a doubt here, Suppose the server has a pre-config of one node and some vnodes of that execution host. Will this API `create_pbsnode` take care of that?

RKORavi replied:
I forgot to check for this condition, but here vnodes will not get created back. The reason is that I am deleting vnodedef file in teardown. I can add the save of vnodedef file in save configurations and load it back in teardown. But i feel that vnodedef file should be under save & load of Mom, since the scope of this PR is limited only to server.
What do you think @sujatapatnaik52 @hirenvadalia @agrawalravi90

hirenvadalia replied:
I am ok with handling vnodedefs with --use-current-setup in separate PR for MoM object

sujatapatnaik52 replied:
I do agree with you. However, we should mention this in the EDD probably, if there are vnodes associated to the nodes then this option is not going load them back.

agrawalravi90 replied:
I'm ok doing this in a different PR as well. I don't think the EDD needs to be modified, an EDD can be implemented via multiple PRs. Maybe instead make a note of it in the PR description?

sujatapatnaik52 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -5127,12 +5127,93 @@ def revert_to_defaults(self, reverthooks=True, revertqueues=True,
+                pass
+            self.manager(MGR_CMD_DELETE, QUEUE, id=queues)
+
+    def delete_sched(self):
```
Do you think will it be more appropriate if we say `delete_sched_config` or maybe something of that sort?

RKORavi replied:
changed

RKORavi commented:
[ptl_run_with_use_current_setup.txt](https://github.com/PBSPro/pbspro/files/3619401/ptl_run_with_use_current_setup.txt)
[printing_configurations.txt](https://github.com/PBSPro/pbspro/files/3619403/printing_configurations.txt)

Test logs with latest changes

agrawalravi90 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -456,6 +457,8 @@ def setUp(self):
         # comm, mom, scheduler
+        if self.use_cur_setup:
+            self.server.delete_nodes()
         if not self.use_cur_setup:
```
make this an else:

RKORavi replied:
changed

agrawalravi90 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -456,6 +457,8 @@ def setUp(self):
         # setup block, rest of them to be added to this block
         # once save & load configurations are implemented for
         # comm, mom, scheduler
+        if self.use_cur_setup:
```
I think you should move the

        if self.use_cur_setup:
            self.server.delete_current_state()
            self.server.load_configuration(self.saved_file)

block here from tearDown() instead of adding this.

RKORavi replied:
I tried to move this block from teardown to setup() but i was still facing the same problem.
So keeping this change

hirenvadalia commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4208,46 +4238,77 @@ def _load_configuration(self, infile, objtype=None):
+                    conf = json.load(f)
+                except ValueError:
+                    self.logger.info("Error loading JSON file: %s"
+                                     % infile)
```
shouldn't we return from here if we fail to load data from file?

hirenvadalia commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4208,46 +4238,77 @@ def _load_configuration(self, infile, objtype=None):
+                    with open(fpath, 'w') as f:
+                        f.write(print_svr)
+                    file_qmgr = open(fpath)
+                    cmd = [qmgr]
```
you don't need cmd here you can pass [qmgr] directly to run_cmd

RKORavi replied:
changed

hirenvadalia commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4208,46 +4238,77 @@ def _load_configuration(self, infile, objtype=None):
+                        f.write(print_svr)
+                    file_qmgr = open(fpath)
+                    cmd = [qmgr]
+                    d = self.du.run_cmd(
```
please add error check here

RKORavi replied:
Hi Hiren,

I am running the command with logerr=False, which will not print anything even if there is a failure. Do you want me log error when exit status is non-zero here?

hirenvadalia replied:
no I mean what will happen if the command fails? shouldn't we stop or do some error handling

RKORavi replied:
Added a check

hirenvadalia commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4208,46 +4238,77 @@ def _load_configuration(self, infile, objtype=None):
+        cmd = [qmgr, "-c", execcmd]
+        self.du.run_cmd(
+            self.hostname, cmd, sudo=True)
+        skip_atb_list = ['id', 'Port', 'pbs_version', 'pcpus',
```
we can't ignore port and state, as in case of multi mom on a single host, the port will be different.

RKORavi replied:
Changed

RKORavi replied:
@hiren 
1. I created multiple moms on a single host and on different port using pbs_config,
pbs_conf --multi-mom --create=4 --init-port=60000
2. Deleted one of the moms & tried to create it back using qmgr command
3. I was getting an error in create node
qmgr -c "c n spartan-6 Port=60006"
qmgr obj=spartan-6 svr=default: Unknown node.

Whereas, when i run the same command for a node on different host it gets created successfully.
qmgr -c "c n 10.75.37.51 Port=60006".
Looks like it will fail in create node and will not go ahead with set node for multi mom on single host.

hirenvadalia replied:
@RKORavi You tagged wrong hiren :) my username is @hirenvadalia

Can you provide me pbs_config output with -lDEBUG when u created multiple moms.

RKORavi replied:
Sorry my bad, corrected it now :)

hirenvadalia replied:
@RKORavi can you please try this commands:
`pbs_config --multi-mom --create=4 --init-port=60000`
`qmgr -c "d n spartan-6"`
`qmgr -c "c n spartan-6 port=60006,Mom=spartan"`

hirenvadalia commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4208,46 +4238,77 @@ def _load_configuration(self, infile, objtype=None):
+            v = str(val)
+            execcmd = "set node %s %s=%s" % (node_name, k, v)
+            cmd = [qmgr, "-c", execcmd]
+            self.du.run_cmd(
```
please add error check here

RKORavi replied:
Added

hirenvadalia commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -5032,59 +5062,26 @@ def revert_to_defaults(self, reverthooks=True, revertqueues=True,
-                    raise
+            self.delete_nodes()
         if reverthooks:
             if self.platform == 'cray' or self.platform == 'craysim':
```
shouldn't this be part of delete_hooks?

RKORavi replied:
For me it looks like setter part in revert to defaults here. I have separated delete configurations from this method

hirenvadalia replied:
but test calls delete_hooks then it will delete these hooks which it shouldn't right... if we don't put this in delete_hooks then test writer has to take care of restoring this hook...

RKORavi replied:
Thanks I got it now.

hirenvadalia commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4995,31 +5049,7 @@ def revert_to_defaults(self, reverthooks=True, revertqueues=True,
                                           'lib', 'python', 'altair',
                                           'pbs_hooks',
                                           'PBS_translate_mpp.HK')
-        self.jacs_hk = os.path.join(self.pbs_conf['PBS_HOME'],
```
you need to restore this as it is used under shasta

RKORavi replied:
I am deleting only site hooks. So, i did not need to add these in delete_hooks() method.
I have changed function name to delete_site_hooks()

RKORavi commented on test/fw/bin/pbs_benchpress:
```diff
@@ -339,6 +341,8 @@ if __name__ == '__main__':
         elif o == '--version':
             print ptl.__version__
             sys.exit(0)
+        elif o == '--use-current-setup':
```
changed

RKORavi commented on test/fw/bin/pbs_benchpress:
```diff
@@ -454,7 +458,8 @@ if __name__ == '__main__':
         loader.set_data(testgroup, testsuites, excludes, follow, testfiles)
         testtags.set_data(tags, eval_tags)
-        runner.set_data(paramfile, testparam, lcov_bin, lcov_data, lcov_out,
+        runner.set_data(paramfile, testparam, use_cur_setup, lcov_bin,
```
changed

RKORavi commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4187,6 +4190,33 @@ def tracejob_match(self, msg, id=None, n=50, tail=True,
+            for m in self.moms:
+                self.moms[m].delete_vnode_defs()
+                self.moms[m].delete_vnodes()
+                self.moms[m].restart()
```
changed

RKORavi commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4187,6 +4190,33 @@ def tracejob_match(self, msg, id=None, n=50, tail=True,
+            self.cleanup_reservations()
+        if del_vdef:
+            for m in self.moms:
+                self.moms[m].delete_vnode_defs()
```
added the check

RKORavi commented:
@agrawalravi90 @hirenvadalia @sujatapatnaik52 should i go head and rebase my branch. I would like to test my PR with recent changes of python3

hirenvadalia commented:
@RKORavi Yes, please go ahead with rebase, it will solve a conflict also in testlib.

agrawalravi90 commented:
@RKORavi No, please give me a chance to review your new changes, you can rebase after you get sign-offs

agrawalravi90 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4246,10 +4227,12 @@ def _load_configuration(self, infile, objtype=None):
                         logerr=False, level=logging.DEBUG)
+                    err_msg = "Failed to load server configurations"
+                    if d['rc'] != 0:
+                        self.logger.error("%s" % err_msg)
```
so, if this (or any of the other object load) fails, what will happen? will pbs_benchpress continue running? Should we stop and error out instead?

RKORavi replied:
Right now, it will continue and will not save the configuration for the failed command.
So, at the time of loading we will just not load any failed commands.

agrawalravi90 replied:
shouldn't we error out if that happens?

hirenvadalia replied:
I agree with @agrawalravi90 we should error out if any command fails instead continue because it won't save correct conf nither it will restore correct.

agrawalravi90 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4263,15 +4246,20 @@ def _load_configuration(self, infile, objtype=None):
-                        self.hostname, cmd, stdin=file_qmgr, sudo=True,
-                        logerr=False, level=logging.DEBUG)
+                        self.hostname, [qmgr], stdin=file_qmgr, sudo=True,
+                        logerr=True, level=logging.DEBUG)
```
logerr is True by default, did you want to set it to False instead?

RKORavi replied:
changed

agrawalravi90 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4283,24 +4271,34 @@ def create_pbsnode(self, node_name, attrs):
+        ret = self.du.run_cmd(self.hostname, cmd, sudo=True)
+        if ret['rc'] != 0:
+            self.logger.info("Failed to create node: %s" % node_name)
+            self.logger.error("Error: %s" % ret['err'])
```
Since you didn't provide logerr=False to run_cmd, it will print out ret['err'] automatically, so please remove this line.

RKORavi replied:
changed

agrawalravi90 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4283,24 +4271,34 @@ def create_pbsnode(self, node_name, attrs):
+                                  logerr=False, level=logging.DEBUG)
+            if ret['rc'] != 0:
+                self.logger.info("Failed to set node attribute %s=%s" % (k, v))
+                self.logger.error("Error: %s" % ret['err'])
```
shouldn't you return False here?

RKORavi replied:
Set node fails even if i try to set Port on mom, but i have to load Port on node to care of condition: multi mom on single host. That is the reason I am not returning False here

RKORavi replied:
Changed

agrawalravi90 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -5248,19 +5250,28 @@ def save_configuration(self, outfile, mode='a'):
+        # pbsnodes -av returns a non zero exit code when there are
+        # no nodes in cluster
+        if ret['rc'] != 0 and err_msg in ret['err']:
+            self.logger.error('Failed to nodes info')
```
type: "failed to get nodes info" instead of "Failed to nodes info"

RKORavi replied:
Changed

agrawalravi90 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -445,17 +450,24 @@ def setUp(self):
+                self.saved_file = path
+            else:
+                self.logger.error("Failed to save test setup")
+                PBSTestSuite.config_saved = True
```
I think we should stop executing and error out if the save fails, what do you think?

RKORavi replied:
Stopping the execution when a command fails.

RKORavi commented:
@hirenvadalia @agrawalravi90 @sujatapatnaik52 please review the latest change

agrawalravi90 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4208,6 +4208,7 @@ def _load_configuration(self, infile, objtype=None):
                         for, one of server, scheduler, mom or
                         if None, load all objects in infile
         """
+        self.unset_svr_attrib()
```
Please move this outside of _load_configuration()

agrawalravi90 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -435,11 +435,12 @@ def setUpClass(cls):
             else:
-                self.logger.error("Failed to save custom setup")
+                cls.logger.error("Failed to save custom setup")
+                sys.exit(1)
```
I think you should do self.fail("Failed to save custom setup") instead of doing sys.exit(1), that way pbs_benchpress can exit gracefully. Same comment elsewhere in this file

hirenvadalia replied:
`self.fail` is only available in test case code, not in this class, you can raise an exception to stop execution flow here

RKORavi commented:
@hirenvadalia Please find the -lDEBUG logs of pbs_config attached
[debug logs.txt](https://github.com/PBSPro/pbspro/files/3653592/debug.logs.txt)

RKORavi commented:
Thanks @hirenvadalia @agrawalravi90 for sign off, rebasing my branch to test with python3 changes

RKORavi commented:
[qmgr server output.txt](https://github.com/PBSPro/pbspro/files/3656947/qmgr.server.output.txt)
[smoketest with option use_current_setup.txt](https://github.com/PBSPro/pbspro/files/3656948/smoketest.with.option.use_current_setup.txt)

RKORavi commented:
Build report:

Machine | OS | Build_Status | Artifacts link | PTL_REPORT | Total_Tests | Passed | Skipped | Failed | Error | Timedout | Build_Start_Time[PDT] | Test_Time | Total_Build_Time
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
pbi-01-r6 | linux26_x86_64 | Success | Artifacts | Results_Link | 1282 | 1272 | 10 | 0 | 0 | 0 | 05:28:17 | 1:16:31 | 01:43:34
pbi-04-r7 | linux310_x86_64 | Success | Artifacts | Results_Link | 1282 | 1272 | 10 | 0 | 0 | 0 | 08:01:06 | 1:23:40 | 01:53:51
pbi-02-r6-altix | linux26_x86_64 | Success | Artifacts | Results_Link | 1282 | 1272 | 10 | 0 | 0 | 0 | 07:50:49 | 1:22:54 | 01:55:15
pbi-11-s11 | linux26_x86_64 | Success | Artifacts | Results_Link | 1282 | 1272 | 10 | 0 | 0 | 0 | 04:34:19 | 0:56:21 | 01:21:36
pbi-18-s12 | linux312_x86_64 | Success | Artifacts | Results_Link | 1282 | 1272 | 10 | 0 | 0 | 0 | 06:15:56 | 0:53:41 | 01:42:47
pbi-23-r7p1-altix | linux310_x86_64 | Success | Artifacts | Results_Link | 1282 | 1272 | 10 | 0 | 0 | 0 | 05:47:07 | 1:18:47 | 01:51:07
pbi-24-s12-altix | linux312_x86_64 | Success | Artifacts | Results_Link | 1282 | 1272 | 10 | 0 | 0 | 0 | 06:35:47 | 0:57:34 | 01:52:12
pbi-25-s11p1-altix | linux26_x86_64 | Success | Artifacts | Results_Link | 1282 | 1272 | 10 | 0 | 0 | 0 | 04:46:37 | 1:08:59 | 01:38:17
pbi-26-s12-cray | linux312_x86_64_crayxt | Success | Artifacts | Results_Link | 1282 | 1271 | 11 | 0 | 0 | 0 | 06:31:41 | 0:56:22 | 01:49:21
pbi-30-s11-cray | linux26_x86_64_crayxt | Success | Artifacts | Results_Link | 1282 | 1271 | 11 | 0 | 0 | 0 | 04:39:09 | 1:01:27 | 01:28:15
pbi-31-u16 | linux413_x86_64 | Success | Artifacts | Results_Link | 1282 | 1272 | 10 | 0 | 0 | 0 | 07:26:31 | 1:10:04 | 02:05:55
pbi-32-s12 | linux312_x86_64_sanitize | Success | Artifacts | Results_Link | 1276 | 1256 | 10 | 0 | 9 | 1 | 11:46:23 | 1:18:59 | 02:12:29
pbi-33-r7 | linux310_x86_64_sanitize | Success | Artifacts | Results_Link | 1282 | 1272 | 10 | 0 | 0 | 0 | 07:09:15 | 1:26:02 | 01:57:31
pbi-34-s15 | linux412_x86_64 | Success | Artifacts | Results_Link | 1282 | 1272 | 10 | 0 | 0 | 0 | 05:48:14 | 1:27:35 | 02:07:44
pbi-35-u18 | linux415_x86_64 | Success | Artifacts | Results_Link | 1282 | 1272 | 10 | 0 | 0 | 0 | 06:57:51 | 1:10:02 | 02:04:23
pbi-36-s15-altix | linux412_x86_64 | Success | Artifacts | Results_Link | 1282 | 1272 | 10 | 0 | 0 | 0 | 05:17:03 | 1:43:59 | 02:13:46
pbi-37-s15-cray | linux412_x86_64_crayxt | Success | Artifacts | Results_Link | 1282 | 1271 | 11 | 0 | 0 | 0 | 04:05:19 | 1:32:20 | 01:58:08
pbi-a01-s12-cray | linux44_aarch64_crayxt | Success | Artifacts | Results_Link | 1282 | 1271 | 11 | 0 | 0 | 0 | 02:33:39 | 2:06:28 | 03:04:57
pbi-a02-s12 | linux44_aarch64.tar.gz | Fail | NA | NA | NA | NA | NA | NA | NA | NA | NA | NA | NA
pbi-a03-r7 | linux45_aarch64 | Success | Artifacts | Results_Link | 1282 | 1272 | 10 | 0 | 0 | 0 | 02:33:36 | 1:08:53 | 01:46:08
pbi-xarm-01-s15 | linux412_aarch64 | Success | Artifacts | Results_Link | 1282 | 1272 | 10 | 0 | 0 | 0 | 05:50:35 | 1:11:43 | 01:43:47
pbi-xarm-02-s15-cray | linux412_aarch64_crayxt | Success | Artifacts | Results_Link | 1282 | 1271 | 11 | 0 | 0 | 0 | 05:39:37 | 1:11:53 | 01:45:34
pbi-38-2k16 | windows | Success | Artifacts |   |   |   |   |   |   |   | 02:34:05 |   | 00:41:30

npadole20 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -5139,6 +5132,87 @@ def revert_to_defaults(self, reverthooks=True, revertqueues=True,
+        Delete sched_priv & sched_log files
+        """
+        self.manager(MGR_CMD_LIST, SCHED)
+        for name in self.schedulers.keys():
```
Please do list(self.schedulers.keys()) here as you are modifying the object at line 5195.

RKORavi replied:
thanks

RKORavi commented:
squashing reviewed commits, so that it will be easier to rebase the branch

RKORavi commented:
[regression results.docx](https://github.com/PBSPro/pbspro/files/3660993/regression.results.docx)

RKORavi commented:
@hirenvadalia @agrawalravi90 please review the changes w.r.t python3 changes, I have attached all the logs.

hirenvadalia commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -457,7 +457,7 @@ def setUpClass(cls):
             else:
                 cls.logger.error("Failed to save custom setup")
-                sys.exit(1)
+                raise Exception("Failed to load custom setup")
```
error message doesn't match with the above line

hirenvadalia commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -474,9 +474,10 @@ def setUp(self):
             else:
                 self.logger.error("Failed to save test setup")
-                sys.exit(1)
+                raise Exception("Failed to load custom setup")
```
same as above

RKORavi commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -4240,7 +4240,9 @@ def _load_configuration(self, infile, objtype=None):
                 # Load pbs.conf file
                 elif k == "pbs_conf":
-                    cfg_vals = base64.decodestring(v)
+                    enc_utf = v.encode('UTF-8')
```
I had to make this change to be compatible with python3. I cannot directly encode a string to base64 in python3. I had to either convert it to byte or encode it in first UTF-8 format. Converting in byte and then trying to use the string proved a little problematic for me. That is the reason i went for encoding it with UTF-8 format & then encoding it in base64 format.
